pkg/upstream/cluster/cluster_manager.go

```diff
@@ -62,12 +62,84 @@ const globalTLSMetrics = "global"
 // types.ClusterManager
 type clusterManager struct {
 	clustersMap      sync.Map
-	protocolConnPool sync.Map // protocolname: { address : connpool }
+	protocolConnPool *connPool
 	tlsMetrics       *mtls.TLSStats
 	tlsMng           atomic.Value // store types.TLSClientContextManager
 	mux              sync.Mutex
 }
 
+type connPool struct {
+	clusterPoolEnable bool     // TODO support modify clusterPoolEnable at runtime
+	globalPool        sync.Map // proto: {addr: pool}
+	clusterPool       sync.Map // proto: {cluster: {addr: pool}}
+}
+
+func newConnPool(clusterPoolEnable bool) *connPool {
+	return &connPool{
+		clusterPoolEnable: clusterPoolEnable,
+	}
+}
+
+func (p *connPool) store(protocolName api.ProtocolName) {
+	p.globalPool.Store(protocolName, &sync.Map{})
+	p.clusterPool.Store(protocolName, &sync.Map{})
+}
+
+func (p *connPool) load(proto types.ProtocolName, snapshot types.ClusterSnapshot) (*sync.Map, bool) {
+	var connectionPool *sync.Map
+	if p.clusterPoolEnable || snapshot.ClusterInfo().IsClusterPoolEnable() {
+		if poolMap, ok := p.clusterPool.Load(proto); ok {
+			value, _ := poolMap.(*sync.Map).LoadOrStore(snapshot.ClusterInfo().Name(), &sync.Map{})
+			connectionPool = value.(*sync.Map)
+			return connectionPool, ok
+		}
+	} else {
+		if poolMap, ok := p.globalPool.Load(proto); ok {
+			connectionPool = poolMap.(*sync.Map)
+			return connectionPool, ok
+		}
+	}
+	return nil, false
+}
+
+func (p *connPool) shutdown(proto types.ProtocolName, addr string) {
+	shutdownPool := func(value interface{}) {
+		connectionPool := value.(*sync.Map)
+		if cp, ok := connectionPool.Load(addr); ok {
+			pool := cp.(types.ConnectionPool)
+			connectionPool.Delete(addr)
+			pool.Shutdown()
+			if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
+				log.DefaultLogger.Debugf("[upstream] [cluster manager] protocol %s address %s connections shutdown", proto, addr)
+			}
+		}
+	}
+	shutdownAll := func(proto types.ProtocolName, addr string) {
+		clusterProtoPool, clusterExists := p.clusterPool.Load(proto)
+		globalProtoPool, globalExists := p.globalPool.Load(proto)
+		if !clusterExists || !globalExists {
+			if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
+				log.DefaultLogger.Debugf("[upstream] [cluster manager] unknown protocol when shutdown, protocol:%s, address: %s", proto, addr)
+			}
+			return
+		}
+		clusterProtoPool.(*sync.Map).Range(func(_, connPool interface{}) bool {
+			shutdownPool(connPool)
+			return true
+		})
+		shutdownPool(globalProtoPool)
+
+	}
+	if proto == "" {
+		p.clusterPool.Range(func(protocol, _ interface{}) bool {
```

Comment:
Indeed, It has a little strange as you say, and maybe we can delete `shutdownAll` to make the logic clearer?
how about this?
```suggestion
	if proto == "" {
		p.clusterPool.Range(func(_, connpool interface{}) bool {
			clusterProtoPool.(*sync.Map).Range(func(_, connPool interface{}) bool {
				shutdownPool(connPool)
				return true
			})
			return true
		})
		p.globalPool.Range(func(_, connpool interface{}) bool {
			shutdownPool(globalProtoPool)
			return true
		}
	} else {
		clusterProtoPool, clusterExists := p.clusterPool.Load(proto)
		globlobalProtoPool, globalExists := p.globalPool.Load(proto)
		if !clusterExists || !globalExists {
			if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
				log.DefaultLogger.Debugf("[upstream] [cluster manager] unknown protocol when shutdown, protocol:%s, address: %s", proto, addr)
			}
			return
		}
		clusterProtoPool.(*sync.Map).Range(func(_, connPool interface{}) bool {
			shutdownPool(connPool)
			return true
		})
		shutdownPool(globalProtoPool)
	}```

Reply:
I agree with it. I fix it later.